<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Document</title>
  </head>
  <body>
    <script>
      //- Được xây dựng dựa trên Promises

      /*1.Async : khai báo một hàm bất đồng bộ (async function someName(){...}).
         - tự động biến hàm thường => promise.
         - khi gọi đến hàm async thì nó sẽ xử lí tất cả và được trả về kết quả trong hàm của nó.
         - Async cho phép sử dụng await.
      */
      //example:
      // từ async trước function => function luôn trả về promise:
      async function f() {
        return 1;
      }
      //sử dụng giống như promise.
      f().then(() => {
        console.log("finished");
      });

      /*
       2.Await: tạm dừng việc thực hiện các hàm async.
         - khi đặt trước một promise, nó sẽ đợi promise và trả về kết quả.
         - await chỉ hoạt động với async, not callback.
         - await chỉ có thể dùng trong function async.

       */
      //cú pháp
      // let value = await promise;
      //example
      // promise giải quyết với 1s.
      async function f2() {
        let promise = new Promise((resolve, reject) => {
          setTimeout(() => {
            resolve("finished f2");
          }, 1000);
        });
        let valuef2 = await promise; // chờ cho đến khi lời hứa được giải quyết
        console.log(valuef2);
      }
      f2();
      /*
      Lưu ý:
      - await chỉ chờ đợi cho đến khi Promise được giải quyết (resolved hoặc rejected).
      promise.then() xử lý kết quả của Promise sau khi nó được giải quyết (resolved).
      - await đợi nhưng không tốn tài nguyên.
      - Await là một cú pháp đơn giản hơn để nhận được kết quả của promise hơn là promise.then.

      - trong promise.them() : xử lí lỗi thông qua việc thêm .catch() trong promise.
      - trong await : xử lí lỗi thông qua cơ chế try / catch.

      -nếu dùng promise.all() xử lí thì 2 tác vụ 5s sẽ chỉ mất 5s thay vì 10s

      -fetch là phương thức call api lấy data.
      */

      // ví dụ cơ chế try / cathch trong await
      async function fetchData() {
        try {
          const response = await fetch(
            "https://jsonplaceholder.typicode.com/users"
          );
          const data = await response.json();
          console.log(data);
        } catch (error) {
          console.error("Lỗi khi lấy dữ liệu:", error);
        }
      }
      fetchData();

      // ví dụ try catch
      try {
        // Chia cho 0 sẽ gây ra lỗi
        console.log(10 / 0);
      } catch (error) {
        console.error("Lỗi:", error.message); // Output: Lỗi: Division by zero
        console.error("Tên lỗi:", error.name); // Output: Tên lỗi: Error
        console.error("Stack trace:", error.stack);
      }
      /*
      Khi một lỗi được ném ra (thrown) trong khối try,
       JavaScript sẽ tạo ra một đối tượng Error chứa các thông tin như:
        error.message: Thông báo lỗi
        error.name: Tên của lỗi
        error.stack: Trace ngược lại stack của code dẫn đến lỗi
      */

      //ví dụ fetch
      console.log(
        "fetch----------------------------------------------------------------------"
      );
      async function getData() {
        await fetch("https://jsonplaceholder.typicode.com/users")
          .then((data) => data.json())
          .then((data) => {
            console.log(data);
          });

        console.log("lấy data đi xử lí");
      }
      getData();

      // console.log("lấy data đi xử lí"); //nó hiện ra trước khi có data

      //promise.all()
      //example
      let p1 = new Promise((resolve, reject) => {
        setTimeout(() => {
          console.log(1);
          resolve();
        }, 2000);
      }).then(() => {
        let p2 = new Promise((resolve, reject) => {
          setTimeout(() => {
            console.log(2);
            resolve();
          }, 3000);
        }).then(() => {
          let p3 = new Promise((resolve, reject) => {
            setTimeout(() => {
              console.log(3);
              resolve();
            }, 3000);
          }).then(() => {
            console.log("finished !");
          });
        });
      });
    </script>
  </body>
</html>
